homeassistant/components/livisi/coordinator.py

```diff
@@ -0,0 +1,135 @@
+"""Code to manage fetching LIVISI data API."""
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Any
+
+from aiohttp import ClientConnectorError
+from aiolivisi import AioLivisi, LivisiEvent, Websocket
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    AVATAR_PORT,
+    CLASSIC_PORT,
+    DEVICE_POLLING_DELAY,
+    LIVISI_REACHABILITY_CHANGE,
+    LIVISI_STATE_CHANGE,
+    LOGGER,
+)
+
+
+class LivisiDataUpdateCoordinator(DataUpdateCoordinator[list[dict[str, Any]]]):
+    """Class to manage fetching LIVISI data API."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, aiolivisi: AioLivisi
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            LOGGER,
+            name="Livisi devices",
+            update_interval=timedelta(seconds=DEVICE_POLLING_DELAY),
+        )
+        self.config_entry = config_entry
+        self.hass = hass
+        self.aiolivisi = aiolivisi
+        self.websocket = Websocket(aiolivisi)
+        self.devices: set[str] = set()
+        self.rooms: dict[str, Any] = {}
+        self.serial_number: str = ""
+        self.controller_type: str = ""
+        self.os_version: str = ""
+        self.is_avatar: bool = False
+        self.port: str = ""
+
+    async def _async_update_data(self) -> list[dict[str, Any]]:
+        """Get device configuration from LIVISI."""
+        try:
+            return await self.async_get_devices()
+        except ClientConnectorError as exc:
+            raise UpdateFailed("Failed to get LIVISI the devices") from exc
+
+    async def async_setup(self) -> None:
+        """Set up the Livisi Smart Home Controller."""
+        if not self.aiolivisi.livisi_connection_data:
+            livisi_connection_data = {
+                "ip_address": self.config_entry.data["host"],
+                "password": self.config_entry.data["password"],
+            }
+
+            await self.aiolivisi.async_set_token(
+                livisi_connection_data=livisi_connection_data
+            )
+        controller_data = await self.aiolivisi.async_get_controller()
+        if controller_data.get("controllerType") == "Avatar":
+            self.port = AVATAR_PORT
+            self.is_avatar = True
+        else:
+            self.port = CLASSIC_PORT
+            self.is_avatar = False
+        self.serial_number = controller_data["serialNumber"]
```

Comment:
Side note for the future: I'd recommend using pydantic in the library to model the data returned by the API and get typed model instances where items can be accessed by attribute.

https://pydantic-docs.helpmanual.io/